java/java-impl-inspections/src/com/intellij/codeInspection/StringRepeatCanBeUsedInspection.java

```diff
@@ -61,12 +69,102 @@ public void visitForStatement(@NotNull PsiForStatement statement) {
+          // Case: textString = textString + " ";
+          leftString = PsiUtil.skipParenthesizedExprDown(concatenation.getOperands()[0]);
+          rightString = concatenation.getOperands()[1];
+        } else if (CONCAT.test(concatCall)) {
+          // Case: textString = textString.concat(" ");
+          leftString = PsiUtil.skipParenthesizedExprDown(concatCall.getMethodExpression().getQualifierExpression());
+          rightString = concatCall.getArgumentList().getExpressions()[0];
+        }
+        if (leftString == null || rightString == null) return null;
+        if (EquivalenceChecker.getCanonicalPsiEquivalence().expressionsAreEquivalent(destinationVariable, leftString)) {
+          // Case: textString = textString + " ";
+          return rightString;
+        } else if (EquivalenceChecker.getCanonicalPsiEquivalence().expressionsAreEquivalent(destinationVariable, rightString)) {
+          // Case: textString = " " + textString;
+          return leftString;
+        }
+        return null;
+      }
+    }
+    if (destinationVariable != null) {
+      // Case: textBuilder = textBuilder.append(" ");
+      appendingExpression = assignmentExpression.getRExpression();
+    }
+    PsiMethodCallExpression call = tryCast(appendingExpression, PsiMethodCallExpression.class);
+    if (call != null) {
+      PsiReferenceExpression caller = tryCast(PsiUtil.skipParenthesizedExprDown(call.getMethodExpression().getQualifierExpression()),
+                                                 PsiReferenceExpression.class);
+      if (caller == null) return null;
+      if (PRINT.test(call)
+          && destinationVariable == null
+          && "out".equals(caller.getReferenceName())) {
+        // Case: System.out.print(" ");
+
+        PsiElement referent = caller.resolve();
+        if (referent instanceof PsiField) {
+          PsiField field = (PsiField)referent;
+          PsiClass containingClass = field.getContainingClass();
+          if (containingClass != null && "java.lang.System".equals(containingClass.getQualifiedName())) {
+            return call.getArgumentList().getExpressions()[0];
+          }
+        }
+      }
+      if (!ExpressionUtil.isEffectivelyUnqualified(caller)) return null;
+      if (ClassUtils.isKnownClassImplementation(caller)) {
+        if (APPEND.test(call)) {
+          // Case: textBuilder.append(" ");
+          return call.getArgumentList().getExpressions()[0];
+        }
+        if (INSERT.test(call)) {
+          // Case: textBuilder.insert(0, " ");
+          if (loop == null || ExplicitArrayFillingInspection.isChangedInLoop(loop, call.getArgumentList().getExpressions()[0])) return null;
+          return call.getArgumentList().getExpressions()[1];
+        }
+        if (destinationVariable != null) return null;
+        if (WRITE.test(call) || WRITE_BYTES.test(call) || PRINT.test(call)) {
```

Comment:
> Can be merged with APPEND case above?

APPEND is not merged because `destinationVariable` can be `null`

> Probably merge CallMatchers together via CallMatcher.anyOf?

Done